### ISSUE TITLE
[css-values] clarify that most (all?) high-dpi devices anchor on pixel, rather than physical, unit

### ISSUE THREAD

**Issue by patrickhlauke** [w3c/csswg-drafts#708]
The section on absolute lengths https://drafts.csswg.org/css-values/#absolute-lengths currently contains the following:

> For print media and similar high-resolution devices, the anchor unit should be one of the standard physical units (inches, centimeters, etc). For lower-resolution devices, and devices with unusual viewing distances, it is recommended instead that the anchor unit be the pixel unit. For such devices it is recommended that the pixel unit refer to the whole number of device pixels that best approximates the reference pixel.

In essence, this splits devices into "print and high resolution" on one side, and "lower-resolution and unusual viewing distances". However, most (all?) high-dpi devices (including recent desktop monitors, laptops, mobile/tablet devices) currently also anchor on the pixel unit. It seems that it's really only print media that (try to) anchor on physical units (unless I'm missing some non-print device that does this?).

I think it would be important to either rewrite this section to change the distinction, or to leave the above as is but then include a note straight after clarifying that currently even high-resolution desktops/laptops/mobiles/tablets are anchored on the pixel unit.

The reason for bringing this up is that this issue seems to constantly rear its head in discussions around being able to use absolute measures like "mm" to actually map to physical size of elements *as rendered on a device* (most recently I had this discussion in the mobile accessibility task force https://www.w3.org/WAI/GL/mobile-a11y-tf/wiki/Summary_of_Research_on_Touch/Pointer_Target_Size#Comparison_of_web_content_sized_in_px_and_physical_size_of_rendered_content_on_common_device_screens which led to having to demonstrate this practically using http://codepen.io/patrickhlauke/full/zqabMR/ ... and I'm aware of similar discussions under WICG in relation to large screen / TV environments)

Happy to draft a PR if this sounds like a reasonable request.

**Comment by nicksherman** [comment]
Related (perhaps duplicate?): Issue #614

**Comment by patrickhlauke** [comment]
related yes. not duplicate though, I'd say (as #614 is proposing a change in what UAs do / new units of measure, while here I'm proposing making the spec match/describe reality a bit more closely)

**Comment by frivoal** [comment]
I agree with this. This isn't really changing how things work, a the way units are anchored to eachother stays the same, and the recommendation to anchor the whole system to either the physical units is only a recommendation, not a requirement. But it is indeed a recommendation that does not match reality, and I don't think it is particularly good advice: what browsers do today is the right thing to do.

**Comment by patrickhlauke** [comment]
@frivoal cool i'll draft a PR in the coming days and fling it over

**Comment by fantasai** [comment]
Wrt the PR here (which, btw, fails to edit css-values-3), it seems to categorize print media with unusual viewing distances into the physical-anchor category, whereas before they were grouped with the pixel-anchor category. If this was an intentional change, I would like a clear resolution on that change from the WG. If that is not an intentional change, then I intend to revert it.

@patrickhlauke can you clarify your intent?

**Comment by patrickhlauke** [comment]
> Wrt the PR here (which, btw, fails to edit css-values-3),

oh, does that mean this PR targetted the wrong branch or file? should i resubmit the PR targetting the right one (which one?)

> it seems to categorize print media with unusual viewing distances into the physical-anchor category, whereas before they were grouped with the pixel-anchor category.

@fantasai as the second part of that section even in the original didn't explicitly say "print", it didn't occur to me here that it even covered "print media with unusual viewing distances" (and i struggle to think of such a device to be honest). But thinking about it now I'd say that once it's print, shouldn't it still be anchored to physical size regardless of what viewing distance the end result is then intended to be viewed at (since the author would presumably account for this, and do so by specifying sizes again with an expectation of them being anchored on actual physical sizes).  Also, to make a simplistic analogy, how would a UA/printer know if something's intended to be viewed at a "usual" or "unusual" distance?

in short, it was not my intention of redefining the anchoring of "print media with unusual viewing distances" when I made the PR, as it was not clear to me that this text implicitly covered it. but i'd say print should always be anchored on physical units regardless

**Comment by fantasai** [comment]
> oh, does that mean this PR targetted the wrong branch or file? should i resubmit the PR targetting the right one (which one?)

No worries, I can port it over. (css-values-3 was forked out, so edits needed to go to both css-values-3/Overview.bs and css-values/Overview.bs)

> it didn't occur to me here that it even covered "print media with unusual viewing distances" (and i struggle to think of such a device to be honest)

Non-electronic billboards and signage? :)

> Also, to make a simplistic analogy, how would a UA/printer know if something's intended to be viewed at a "usual" or "unusual" distance?

On the one hand, there's that. And on the other, would you want electronic and non-electronic signage to behave the same? There's a *huge* gulf between 1/96in in physical units and 1px in angular units at far distances.

Anyway, I'll Agenda+ to sort this out then.

**Comment by frivoal** [comment]
My take on this is that we shouldn't obsess to much, since this is only a should. As print media with unusual viewing distances are not the typical case, it's perfectly fine to deviate from the recommendation when you have better knowledge of what this is actually for.

**Comment by patrickhlauke** [comment]
> Non-electronic billboards and signage?

ah, good point.

> wouldn't you want electronic and non-electronic signage to behave the same?

while i can see both sides of this, my (purely personal of course) view here is that for a designer, when they work on paper/print, they have a stronger expectation that setting something in physical units will also match real-world physical units, whereas when designing for screen (knowing that screens have all sorts of variables like resolution etc) there's perhaps more of an expectation that everything anchors to the CSS pixel (which ideally should be close to the reference pixel, but even that varies across devices/OSs/UAs). also, purely empirically, i don't know many designers who use physical units like `mm` when working for screen (though i know many designers who "discover" these physical units and are then quite disappointed that on screen it's anchored on pixel and usually they don't match up to real-world units as measured on the physical screen, as per https://www.w3.org/WAI/GL/mobile-a11y-tf/wiki/Summary_of_Research_on_Touch/Pointer_Target_Size#Comparison_of_web_content_sized_in_px_and_physical_size_of_rendered_content_on_common_device_screens)

**Comment by patrickhlauke** [comment]
> we shouldn't obsess to much, since this is only a should

that's also a sensible take / get-out-of-jail card :)

**Comment by patrickhlauke** [comment]
so if a physical printer is deemed "low resolution", using physical CSS units will actually have them anchored on px rather than trying to reflect their real physical size? and same for large scale print? can't say i'm exactly thrilled by this...so a printer will use physical units only if it falls between "not low resolution, not aimed at large print"?

**Comment by patrickhlauke** [comment]
@fantasai also, simply moving the parenthesis won't be enough, because otherwise now you have two conflicting sentences:

> For print media, the anchor unit should be one of the standard physical units (inches, centimeters, etc).

which would imply ALL print media, followed by

>For screen media (including high-resolution devices), lower-resolution devices, and devices with unusual viewing distances, it is recommended instead that the anchor unit be the pixel unit.

which, in light of the preceding sentence, would suggest that it doesn't apply to print media regardless of resolution or viewing distance

**Comment by fantasai** [comment]
Okay, checked in some wording tweaks' per @plinss’s advice. Let me know if this is sufficient. Keep in mind these are recommendations: a UA can do something different if it has a reason for it.

**Comment by patrickhlauke** [comment]
While I still think that print should always be anchored on physical sizes regardless of viewing distance etc, I agree the rewritten section are now clearer without conflict. Thanks @fantasai and @plinns

**Comment by patrickhlauke** [comment]
Closing, as my primary concern (to specify that even high-resolution screen media are still anchored on pixels, to match reality) is addressed.

**Comment by Crissov** [comment]
It makes no sense whatsoever that browsers on highly integrated devices like phones, tablets and TVs with known fixed screen dimension and resolution do not anchor on physical units.

**Comment by patrickhlauke** [comment]
Taking just one OS as an example, does Android provide exact measurement in physical units of the screen as an API/value that can be queried by native apps? as far as I'm aware, a native app can query "Screen Size", but the return values from the OS are only "small, normal, large, and extra-large".

Assume now a browser gets downloaded from Google Play and installed (so it's not preinstalled by the manufacturer/OEM) . How can the browser know what the actual physical dimension of the screen of that particular device is?

**Comment by Crissov** [comment]
I was indeed thinking of highly integrated devices, e.g. iOS ones and TVs (not STBs), although Android *should* be able too provide reliable information to apps as well.
